plugins/nexusoperations/statemachine.go

```diff
@@ -0,0 +1,556 @@
+					ApplicationFailureInfo: &failurepb.ApplicationFailureInfo{
+						NonRetryable: true,
+					},
+				},
+			}
+		}
+		// Keep last attempt information as-is for debuggability when completed asynchronously.
+		// When used in a workflow, this machine node will be deleted from the tree after this transition.
+		return op.output(event.Node)
+	},
+)
+
+// EventSucceeded is triggered when an invocation attempt succeeds.
+type EventSucceeded struct {
+	// Only set if the operation completed synchronously, as a response to a StartOperation RPC.
+	AttemptTime *time.Time
+	Node        *hsm.Node
+}
+
+var TransitionSucceeded = hsm.NewTransition(
+	[]enumsspb.NexusOperationState{
+		enumsspb.NEXUS_OPERATION_STATE_SCHEDULED,
+		enumsspb.NEXUS_OPERATION_STATE_BACKING_OFF,
+		enumsspb.NEXUS_OPERATION_STATE_STARTED,
+	},
+	enumsspb.NEXUS_OPERATION_STATE_SUCCEEDED,
+	func(op Operation, event EventSucceeded) (hsm.TransitionOutput, error) {
+		if event.AttemptTime != nil {
+			op.recordAttempt(*event.AttemptTime)
+		}
+		// Keep last attempt information as-is for debuggability when completed asynchronously.
+		// When used in a workflow, this machine node will be deleted from the tree after this transition.
+		return op.output(event.Node)
+	},
+)
+
+// EventCanceled is triggered when an invocation attempt succeeds.
+type EventCanceled struct {
+	// Only set if the operation completed synchronously, as a response to a StartOperation RPC.
+	AttemptFailure *AttemptFailure
+	Node           *hsm.Node
+}
+
+var TransitionCanceled = hsm.NewTransition(
+	[]enumsspb.NexusOperationState{
+		enumsspb.NEXUS_OPERATION_STATE_SCHEDULED,
+		enumsspb.NEXUS_OPERATION_STATE_BACKING_OFF,
+		enumsspb.NEXUS_OPERATION_STATE_STARTED,
+	},
+	enumsspb.NEXUS_OPERATION_STATE_CANCELED,
+	func(op Operation, event EventCanceled) (hsm.TransitionOutput, error) {
+		if event.AttemptFailure != nil {
+			op.recordAttempt(event.AttemptFailure.Time)
+			op.LastAttemptFailure = &failurepb.Failure{
+				Message: event.AttemptFailure.Err.Error(),
+				FailureInfo: &failurepb.Failure_CanceledFailureInfo{
+					CanceledFailureInfo: &failurepb.CanceledFailureInfo{},
+				},
+			}
+		}
+		// Keep last attempt information as-is for debuggability when completed asynchronously.
+		// When used in a workflow, this machine node will be deleted from the tree after this transition.
+		return op.output(event.Node)
+	},
+)
+
+// EventStarted is triggered when an invocation attempt succeeds and the handler indicates that it started an
+// asynchronous operation.
+type EventStarted struct {
+	Time       time.Time
+	Node       *hsm.Node
+	Attributes *historypb.NexusOperationStartedEventAttributes
+}
+
+var TransitionStarted = hsm.NewTransition(
+	[]enumsspb.NexusOperationState{enumsspb.NEXUS_OPERATION_STATE_SCHEDULED},
+	enumsspb.NEXUS_OPERATION_STATE_STARTED,
+	func(op Operation, event EventStarted) (hsm.TransitionOutput, error) {
+		op.recordAttempt(event.Time)
```

Comment:
We don't. Where did you see that?

Reply:
Inside `recordAttempt`, which is called during the "start" transition right here, and is also called in the "succeeded" and "failed" transitions.

Reply:
This isn't when the transition starts and finishes. It's when an start operation request completes synchronously or with an operation ID indicating that it will conclude asynchronously - using the provided callback.